Rework episode 02 - cut down on Git content (#198)
carpentries-incubator/fair-research-software: episodes/02-version-control.md

If you are not already inside this directory, from your command line terminal you can navigate to it and list its 
contents with:

```bash
cd ~/spacewalks
ls -la
total 272
drwx------@  4 mbassan2  staff     128 28 May 20:06 .
drwxr-x---+ 66 mbassan2  staff    2112 28 May 20:07 ..
-rw-rw-r--@  1 mbassan2  staff  132981  4 Apr 10:48 data.json
-rw-rw-r--@  1 mbassan2  staff    1518  4 Apr 10:48 my code v2.py
```

Over the rest of the course, we will transform a collection of these files into a well-structured software project that 
follows established good practices in research software engineering.

## Version control

Before we start making changes to our code, let's make sure we can keep a history of what changes we have done since 
we inherited the code from our colleague.

We can track changes with version control. Later on, we will store those changes on a remote server too --
both for safe-keeping and to make them easier to share with others. In later episodes, we will also see how version control 
makes it easier for multiple collaborators to work together on the same project at the same time and combine 
their contributions.

:::::: callout 

> ### What is a version control system?

Reviewer: ```suggestion
### Version control refresher

#### What is a version control system?
```

Version control systems are tools that let you track changes in files over time in a special database that allows 
users to "travel through time", and compare earlier versions of the files with the current state.
Think of a version control system like turning on 'Track Changes' on Microsoft Word/Google Docs,
but for *any* files you want, and a lot more powerful and flexible.

> **Why use a version control system?**

Reviewer: ```suggestion
#### Why use a version control system?
```

we can restore our project to the state it was at any point in time.
This is incredibly useful for reproducing results from a specific version of the code,
or tracking down which changes we (or others) made introduced bugs or changed our results.

The other benefit of version control is it provides us with a **history** of our development.
As we make each change, we record what it was, and why we made it.
This helps make our development process transparent and auditable -- which is a good scientific practice.

It also makes our project more **sustainable** - as our data, software and methods (knowledge) remain usable and
accessible over time (especially if made available in shared version controlled code repositories), even after the original funding ends or team members move on.

::::::

### Git version control system

**Git** is the most popular version control system used by researchers worldwide, and the one we'll be using.
Git is used mostly for managing code when developing software, but it can track *any* files --
and is particularly effective with text-based files (e.g. source code like `.py`, `.c`, `.r`, but also `.csv`, `.tex` and more).

> Git allows multiple people to work on the same project (even the same file) at the same time.

Reviewer: ```suggestion
Git helps multiple people work on the same project (even the same file) at the same time.
```

  Best practice when collaborating with others on a shared repository is to always do a `git pull` before a `git push`,
  to ensure you have any latest changes before you push your own.

::::::

> ### Create a new repository

Reviewer: ```suggestion
### Start tracking changes with Git
```

nothing added to commit but untracked files present (use "git add" to track)
```

This tells us that Git has noticed two files in our directory, but unlike Dropbox or OneDrive,
it does not *automatically* track them. We need to tell Git explicitly which files we want it to track.
This is not a handicap, but rather helpful, since scientific code can have vast inputs or outputs we might not want 
Git to track and store (GBs ot TBs of space telescope data) or require sensitive information we cannot share
(for example, medical records).

### Add files into repository

> We can tell Git to track a file using `git add`:

Reviewer: Should we demonstrate how to do these version control steps with the VS Code graphical interface? Or do you prefer to keep people working in the terminal as much as possible?

Author: We have a separate issue for this now.

$ git status
```

`git mv <old> <new>` command is the equivalent of running `mv <old> <new>` immediately followed by `git add <new>`,

> so the changes have been staged automatically. We just need to commit them.

Reviewer: I think there's another advantage here: with `mv` only, Git thinks one file has been deleted and a new one created. Whereas, when you use `git mv`, Git "understands" that it's the same file but with a new name.

Reviewer: ```suggestion
`git mv <old> <new>` command is the equivalent of running `mv <old> <new>` immediately followed by `git add <old> <new>`,
so the changes have been staged automatically. We just need to commit them.
```

A small change that gets the same point across that I was trying to make.
